File: backend/app/api/v1/residents.py
"""Residents endpoints — signup, Admin approval, property linking."""
import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.db import get_db
from app.core.security import CurrentUser, require_role
from app.models.enums import Role, UserStatus
from app.schemas.resident import (
    AdminSelfResidentLinkIn,
    OwnerContactUpdateIn,
    PropertyLinkRequestDecisionIn,
    PropertyLinkRequestIn,
    PropertyLinkRequestOut,
    PropertyResidentLinkIn,
    PropertyResidentOut,
    ResidentApprovalIn,
    ResidentOut,
    ResidentSignupIn,
)
from app.services import resident_service
from app.services.scope_service import subadmin_has_scope_over_property

router = APIRouter(prefix="/residents", tags=["residents"])


@router.post("/signup", response_model=ResidentOut)
async def signup(
    body: ResidentSignupIn, db: Annotated[AsyncSession, Depends(get_db)]
) -> ResidentOut:
    """Public — no auth required. Resident waits for Admin approval
    (Section 26)."""
    resident = await resident_service.signup_resident(db, body)
    return ResidentOut.model_validate(resident)


@router.get("", response_model=list[ResidentOut])
async def list_all(
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.ADMIN))],
    status_filter: Annotated[UserStatus | None, Query(alias="status")] = None,
) -> list[ResidentOut]:
    """Admin's resident directory — e.g. the "Make Sub-admin" picker on
    the Residents page (?status=ACTIVE), separate from list_pending below
    which is specifically the approval queue."""
    residents = await resident_service.list_residents(db, current.society_id, status_filter)
    return [ResidentOut.model_validate(r) for r in residents]


@router.get("/pending", response_model=list[ResidentOut])
async def list_pending(
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.ADMIN))],
) -> list[ResidentOut]:
    residents = await resident_service.list_pending_residents(db, current.society_id)
    return [ResidentOut.model_validate(r) for r in residents]


@router.post("/{resident_id}/approval", response_model=ResidentOut)
async def decide_approval(
    resident_id: uuid.UUID,
    body: ResidentApprovalIn,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.ADMIN))],
) -> ResidentOut:
    resident = await resident_service.decide_resident_approval(
        db, current.society_id, resident_id, body.approve, current.user_id
    )
    return ResidentOut.model_validate(resident)


@router.post("/self-link", response_model=PropertyResidentOut)
async def link_self(
    body: AdminSelfResidentLinkIn,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.ADMIN))],
) -> PropertyResidentOut:
    """Admin links themselves (not another user) to a property in their own
    society as Owner/Tenant, gaining a RESIDENT role on their existing
    account — no separate approval needed, since the caller already runs
    the society. See resident_service.link_admin_as_resident."""
    link = await resident_service.link_admin_as_resident(db, current.society_id, current.user_id, body)
    return PropertyResidentOut.model_validate(link)


@router.post("/property-links", response_model=PropertyResidentOut)
async def link_property(
    body: PropertyResidentLinkIn,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.ADMIN))],
) -> PropertyResidentOut:
    link = await resident_service.link_resident_to_property(db, current.society_id, body)
    return PropertyResidentOut.model_validate(link)


@router.delete("/property-links/{link_id}", response_model=PropertyResidentOut)
async def unlink_property(
    link_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.ADMIN))],
) -> PropertyResidentOut:
    link = await resident_service.unlink_resident_from_property(db, current.society_id, link_id)
    return PropertyResidentOut.model_validate(link)


@router.patch("/property-links/{link_id}/owner-contact", response_model=PropertyResidentOut)
async def update_owner_contact(
    link_id: uuid.UUID,
    body: OwnerContactUpdateIn,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.RESIDENT))],
) -> PropertyResidentOut:
    """A Tenant self-recording the Owner's contact details on their own
    Profile page — see resident_service.update_owner_contact's docstring
    for why this exists (a Tenant can sign up with no Owner account in
    the system at all)."""
    link = await resident_service.update_owner_contact(db, current.society_id, current.user_id, link_id, body)
    return PropertyResidentOut.model_validate(link)


@router.get("/by-property/{property_id}", response_model=list[PropertyResidentOut])
async def get_property_residents(
    property_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.ADMIN, Role.SUB_ADMIN))],
) -> list[PropertyResidentOut]:
    # Audit fix: a Sub-admin could previously name ANY property in their own
    # society and see its occupants — Wing/Row scope (Section 27) wasn't
    # applied here. Admin has no such restriction.
    if current.active_role == Role.SUB_ADMIN and not await subadmin_has_scope_over_property(
        db, current.user_id, property_id, current.society_id
    ):
        raise HTTPException(status.HTTP_403_FORBIDDEN, "This property is outside your assigned scope")
    links = await resident_service.list_property_residents(db, current.society_id, property_id)
    return [PropertyResidentOut.model_validate(link) for link in links]


@router.post("/property-link-requests", response_model=PropertyLinkRequestOut)
async def request_property_link(
    body: PropertyLinkRequestIn,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.RESIDENT))],
) -> PropertyLinkRequestOut:
    """An already-ACTIVE Resident requesting (from their own Profile page)
    to link themselves to an additional property — stays PENDING until
    the Admin approves it (Section 12/26)."""
    req = await resident_service.submit_property_link_request(db, current.society_id, current.user_id, body)
    return PropertyLinkRequestOut.model_validate(req)


@router.get("/property-link-requests/mine", response_model=list[PropertyLinkRequestOut])
async def my_property_link_requests(
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.RESIDENT))],
) -> list[PropertyLinkRequestOut]:
    reqs = await resident_service.list_my_property_link_requests(db, current.society_id, current.user_id)
    return [PropertyLinkRequestOut.model_validate(r) for r in reqs]


@router.get("/property-link-requests/pending", response_model=list[PropertyLinkRequestOut])
async def pending_property_link_requests(
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.ADMIN))],
) -> list[PropertyLinkRequestOut]:
    reqs = await resident_service.list_pending_property_link_requests(db, current.society_id)
    return [PropertyLinkRequestOut.model_validate(r) for r in reqs]


@router.post("/property-link-requests/{request_id}/decision", response_model=PropertyLinkRequestOut)
async def decide_property_link_request(
    request_id: uuid.UUID,
    body: PropertyLinkRequestDecisionIn,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.ADMIN))],
) -> PropertyLinkRequestOut:
    req = await resident_service.decide_property_link_request(
        db, current.society_id, request_id, body.approve, body.decision_reason, current.user_id
    )
    return PropertyLinkRequestOut.model_validate(req)


@router.get("/{resident_id}/properties", response_model=list[PropertyResidentOut])
async def get_resident_properties(
    resident_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.ADMIN, Role.SUB_ADMIN, Role.RESIDENT))],
) -> list[PropertyResidentOut]:
    # A Resident may only query their own properties — not another resident's
    # (private data boundary; Admin has no such restriction here).
    if current.active_role == Role.RESIDENT and resident_id != current.user_id:
        raise HTTPException(status.HTTP_403_FORBIDDEN, "Residents can only view their own properties")
    links = await resident_service.list_resident_properties(db, current.society_id, resident_id)
    # Audit fix: a Sub-admin could previously query ANY resident's full
    # property list regardless of scope. A resident can be linked to
    # properties across different wings, so filter per-link, not per-request.
    if current.active_role == Role.SUB_ADMIN:
        links = [
            link for link in links
            if await subadmin_has_scope_over_property(db, current.user_id, link.property_id, current.society_id)
        ]
    return [PropertyResidentOut.model_validate(link) for link in links]

File: frontend/src/api/residents.ts
import { apiClient } from "@/api/client";
import type { LocationType, RelationshipType, Role, RoleRequestStatus, UserStatus } from "@/types/enums";

export interface ResidentOut {
  id: string;
  society_id: string | null;
  full_name: string;
  mobile: string;
  email: string | null;
  status: UserStatus;
  created_at: string;
}

export interface PropertyResidentOut {
  id: string;
  property_id: string;
  resident_id: string;
  relationship_type: RelationshipType;
  is_active: boolean;
  start_date: string | null;
  end_date: string | null;
  owner_contact_name: string | null;
  owner_contact_mobile: string | null;
}

export interface SubAdminScopeOut {
  id: string;
  sub_admin_id: string;
  location_id: string;
  assigned_by: string;
  assigned_at: string;
  revoked_at: string | null;
}

export interface SubAdminAssignmentOut {
  scope_id: string;
  sub_admin_id: string;
  sub_admin_name: string;
  sub_admin_mobile: string;
  location_id: string;
  location_name: string;
  location_type: LocationType;
  assigned_at: string;
}

export interface RoleRequestOut {
  id: string;
  user_id: string;
  society_id: string;
  status: RoleRequestStatus;
  reason: string | null;
  reviewed_by: string | null;
  reviewed_at: string | null;
  decision_reason: string | null;
  created_at: string;
}

export interface PropertyLinkRequestOut {
  id: string;
  society_id: string;
  resident_id: string;
  property_id: string;
  relationship_type: RelationshipType;
  status: RoleRequestStatus;
  reason: string | null;
  reviewed_by: string | null;
  reviewed_at: string | null;
  decision_reason: string | null;
  created_at: string;
}

export const residentsApi = {
  /** Public — no auth required. Resident waits for their society's Admin
   * to approve (see pending/decideApproval below). property_id/
   * relationship_type are required — the Resident picks their own house
   * (from societiesApi.publicProperties) and Owner/Tenant right at
   * signup, same for a Flats or Bungalow society; the link exists
   * immediately but stays inert until Admin approval. */
  signup: (input: {
    full_name: string; mobile: string; email?: string; society_id: string;
    property_id: string; relationship_type: RelationshipType;
  }) => apiClient.post<ResidentOut>("/residents/signup", input),
  /** Admin's resident directory — e.g. the "Make Sub-admin" picker
   * (status="ACTIVE"), separate from pending() below which is
   * specifically the approval queue. Omit status for every resident
   * regardless of approval state. */
  list: (status?: UserStatus) => apiClient.get<ResidentOut[]>("/residents", { params: { status } }),
  pending: () => apiClient.get<ResidentOut[]>("/residents/pending"),
  decideApproval: (id: string, approve: boolean, rejectionReason?: string) =>
    apiClient.post<ResidentOut>(`/residents/${id}/approval`, { approve, rejection_reason: rejectionReason }),
  linkProperty: (propertyId: string, residentId: string, relationshipType: RelationshipType) =>
    apiClient.post<PropertyResidentOut>("/residents/property-links", {
      property_id: propertyId, resident_id: residentId, relationship_type: relationshipType,
    }),
  unlinkProperty: (linkId: string) => apiClient.delete<PropertyResidentOut>(`/residents/property-links/${linkId}`),
  /** Admin-only — links the CALLER (not another user) to a property in
   * their own society, granting a RESIDENT role on their existing account
   * if they don't already have one (Section 4 dual-role: ADMIN+RESIDENT).
   * No approval step — the caller already runs the society. */
  linkSelf: (propertyId: string, relationshipType: RelationshipType) =>
    apiClient.post<PropertyResidentOut>("/residents/self-link", {
      property_id: propertyId, relationship_type: relationshipType,
    }),
  byProperty: (propertyId: string) => apiClient.get<PropertyResidentOut[]>(`/residents/by-property/${propertyId}`),
  /** Powers the Assign Sub-admin page's picker — Admin picks a Wing/Row
   * first, then one of the residents actually living there. */
  byLocation: (locationId: string) => apiClient.get<ResidentOut[]>(`/residents/by-location/${locationId}`),
  /** Tenant-only, on their own active link — self-recording the Owner's
   * contact details (free text, not a real account) since a Tenant can
   * now sign up with no Owner account in the system to look this up
   * from. Pass null/omit to clear a field. */
  updateOwnerContact: (linkId: string, ownerContactName?: string | null, ownerContactMobile?: string | null) =>
    apiClient.patch<PropertyResidentOut>(`/residents/property-links/${linkId}/owner-contact`, {
      owner_contact_name: ownerContactName, owner_contact_mobile: ownerContactMobile,
    }),
  /** Resident-only — from their own Profile page, requesting to link
   * themselves to an additional property. Unlike linkSelf/signup, this
   * stays PENDING until the Admin approves it (see decidePropertyLinkRequest
   * below) — approval is what actually creates the real link. */
  requestPropertyLink: (propertyId: string, relationshipType: RelationshipType, reason?: string) =>
    apiClient.post<PropertyLinkRequestOut>("/residents/property-link-requests", {
      property_id: propertyId, relationship_type: relationshipType, reason,
    }),
  /** Resident's own requests, any status — so their Profile page can show
   * pending/approved/rejected instead of the request just vanishing. */
  myPropertyLinkRequests: () => apiClient.get<PropertyLinkRequestOut[]>("/residents/property-link-requests/mine"),
  pendingPropertyLinkRequests: () =>
    apiClient.get<PropertyLinkRequestOut[]>("/residents/property-link-requests/pending"),
  decidePropertyLinkRequest: (requestId: string, approve: boolean, decisionReason?: string) =>
    apiClient.post<PropertyLinkRequestOut>(`/residents/property-link-requests/${requestId}/decision`, {
      approve, decision_reason: decisionReason,
    }),
};

export const subadminsApi = {
  /** Every active Sub-admin scope in the society, denormalized with
   * names — the Assign Sub-admin page's "Current Sub-admins" overview. */
  listAll: () => apiClient.get<SubAdminAssignmentOut[]>("/subadmins"),
  promote: (residentId: string, locationIds: string[]) =>
    apiClient.post<SubAdminScopeOut[]>("/subadmins/promote", { resident_id: residentId, location_ids: locationIds }),
  scopes: (subAdminId: string) => apiClient.get<SubAdminScopeOut[]>(`/subadmins/${subAdminId}/scopes`),
  assignScope: (subAdminId: string, locationId: string) =>
    apiClient.post<SubAdminScopeOut>(`/subadmins/${subAdminId}/scopes`, { location_id: locationId }),
  revokeScope: (scopeId: string) => apiClient.delete<SubAdminScopeOut>(`/subadmins/scopes/${scopeId}`),
  /** Admin directly removes someone's Sub-admin role and every active
   * scope, in one shot — no resignation request needed. */
  demote: (subAdminId: string) => apiClient.delete(`/subadmins/${subAdminId}`),
  submitResignation: (reason?: string) => apiClient.post<RoleRequestOut>("/subadmins/resignation", { reason }),
  pendingResignations: () => apiClient.get<RoleRequestOut[]>("/subadmins/resignations/pending"),
  decideResignation: (requestId: string, approve: boolean, decisionReason?: string) =>
    apiClient.post<RoleRequestOut>(`/subadmins/resignations/${requestId}/decision`, { approve, decision_reason: decisionReason }),
};

export type { Role };
